kasa/klapprotocol.py

```diff
@@ -0,0 +1,155 @@
+        self.authenticator = authentication.authenticator()
+        self.handshake_lock = asyncio.Lock()
+        self.handshake_done = False
+
+        super().__init__(host=host)
+
+        _LOGGER.debug("[KLAP] Created KLAP object for %s", self.host)
+
+    @staticmethod
+    def _sha256(payload: bytes) -> bytes:
+        return hashlib.sha256(payload).digest()
+
+    async def _handshake(self, session) -> None:
+        _LOGGER.debug("[KLAP] Starting handshake with %s", self.host)
+
+        # Handshake 1 has a payload of client_challenge
+        # and a response of 16 bytes, followed by sha256(clientBytes | authenticator)
+
+        url = f"http://{self.host}/app/handshake1"
+        resp = await session.post(url, data=self.client_challenge)
+        _LOGGER.debug("Got response of %d to handshake1", resp.status)
+        if resp.status != 200:
+            raise SmartDeviceException(
+                "Device responded with %d to handshake1" % resp.status
+            )
+
+        response = await resp.read()
+        self.server_challenge = response[0:16]
+        server_hash = response[16:]
+
+        _LOGGER.debug("Server bytes are: %s", self.server_challenge.hex())
+        _LOGGER.debug("Server hash is: %s", server_hash.hex())
+
+        # Check the response from the device
+        local_hash = self._sha256(self.client_challenge + self.authenticator)
+
+        if local_hash != server_hash:
+            _LOGGER.debug(
+                "Expected %s got %s in handshake1",
+                local_hash.hex(),
+                server_hash.hex(),
+            )
+            raise SmartDeviceException("Server response doesn't match our challenge")
+        else:
+            _LOGGER.debug("handshake1 hashes match")
+
+        # We need to include only the TP_SESSIONID cookie - aiohttp's cookie handling
+        # adds a bogus TIMEOUT cookie
+        cookie = session.cookie_jar.filter_cookies(url).get("TP_SESSIONID")
+        session.cookie_jar.clear()
+        session.cookie_jar.update_cookies({"TP_SESSIONID": cookie}, URL(url))
+        _LOGGER.debug("Cookie is %s", cookie)
+
+        # Handshake 2 has the following payload:
+        #    sha256(serverBytes | authenticator)
+        url = f"http://{self.host}/app/handshake2"
+        payload = self._sha256(self.server_challenge + self.authenticator)
+        resp = await session.post(url, data=payload)
+        _LOGGER.debug("Got response of %d to handshake2", resp.status)
+        if resp.status != 200:
+            raise SmartDeviceException(
+                "Device responded with %d to handshake2" % resp.status
+            )
+
+        # Done handshaking, now we need to compute the encryption keys
+        agreed = self.client_challenge + self.server_challenge + self.authenticator
+        self.encrypt_key = self._sha256(b"lsk" + agreed)[:16]
+        self.hmac_key = self._sha256(b"ldk" + agreed)[:28]
+        fulliv = self._sha256(b"iv" + agreed)
```

Comment:
Group local & instance assignments separately, consider adding some newlines to make it nicer to read.